### PR TITLE
Write SMG for using PixelData

### PR DESCRIPTION
**Pull request by hsautessella**
I've written an SMG for using the `PixelData` class. I've written some context on the problem it aims to address, as well as some principles behind the object. There are also a couple of examples on how to implement file-backed algorithms, and a few notes on things to be aware of.

Addresses #618 

_Note: I expect there will be some additions following our meeting tomorrow (24/03/21)_

### CODE REVIEW

File: documentation/smg/17_using_pixel_data.md
Comment (oerc0122) [inline documentation/smg/17_using_pixel_data.md:1]: I would probably say this is more "Pixel Data" than just using it. Especially given you start with a definition of why it's needed.

File: documentation/smg/17_using_pixel_data.md
Comment (oerc0122) [inline documentation/smg/17_using_pixel_data.md:3]: Problems?

File: documentation/smg/17_using_pixel_data.md
Comment (oerc0122) [inline documentation/smg/17_using_pixel_data.md:81]: Is there a method to set page sizes such that each page is of such a size the data can be evenly distributed between N workers or similar? Might save some work
```matlab
pixData.page_size = pixData.distribute(nWorkers, maxPage)
```
Such that `page_size` will be some multiple of `nWorkers` and no bigger than `maxPage`

File: documentation/smg/17_using_pixel_data.md
Comment (oerc0122) [inline documentation/smg/17_using_pixel_data.md:110]: Are they kept as temporaries or are they made part of the core? Does it still refer to the original SQW?

File: documentation/smg/17_using_pixel_data.md
Comment (oerc0122) [inline documentation/smg/17_using_pixel_data.md:111]: This should in principle be possible using a doubly dirty notation? Copy the dirties as 1, if they're subsequently dirtied, they're set to 2 and new tmps are created. On copy set all 2s to 1.

i.e. 0 -> SQW, 1 -> Temps, 2-> MyTemps, but then I guess when you chain copy, this gets awkward?

File: documentation/smg/17_using_pixel_data.md
Comment (oerc0122) [inline documentation/smg/17_using_pixel_data.md:155]: Is it an error to fail to do so?

File: documentation/smg/17_using_pixel_data.md
Comment (oerc0122) [inline documentation/smg/17_using_pixel_data.md:190]: can

File: documentation/smg/17_using_pixel_data.md
Comment (oerc0122) [inline documentation/smg/17_using_pixel_data.md:222]: Provided they are contiguous

File: documentation/smg/17_using_pixel_data.md
Comment (oerc0122) [inline documentation/smg/17_using_pixel_data.md:299]: Does it always mark it as dirty, even if no changes have taken place? Is it possible to set the dirty marking to any setters/getters?

File: documentation/smg/17_using_pixel_data.md
Comment (oerc0122) [inline documentation/smg/17_using_pixel_data.md:323]: a `dnd` object

File: documentation/smg/17_using_pixel_data.md
Comment (hsautessella) [reply documentation/smg/17_using_pixel_data.md:81]: No such method as yet. Changing page size once the object already exists is quite fiddly; what happens if there are dirty pages stored in temp files? For the time being, page size can only be set on construction.

Comment (oerc0122) [reply documentation/smg/17_using_pixel_data.md:81]: I also realised that if you were keen on doing this, you could set it via `split_vector_into_chunks` (or whatever it's called)

File: documentation/smg/17_using_pixel_data.md
Comment (hsautessella) [reply documentation/smg/17_using_pixel_data.md:111]: We did consider this, but thought that it would get a bit messy. Copying edited file-backed objects is not something we have done anywhere at the moment. If the need arises, we could look at optimizing this, but it may be premature to do so at this point.

File: documentation/smg/17_using_pixel_data.md
Comment (hsautessella) [reply documentation/smg/17_using_pixel_data.md:155]: The values you set must be numeric, so it would be an error to pass any of the setters anything other than an in-memory array.

Although, no `set_pixels` method exists at the moment (`set_data` was only recently added), but that could in theory set pixel data from a file-backed `PixelData` object.

Should I remove the setters part of this in case a `set_pixels` method is ever required?

Comment (oerc0122) [reply documentation/smg/17_using_pixel_data.md:155]: I think the getters bit is clarified below but useful. I just think it would be good to note what happens if you attempt to set pixels out of current page, or set pixels out of range? If it's an error, note "Anything else will result in an error"

File: documentation/smg/17_using_pixel_data.md
Comment (hsautessella) [reply documentation/smg/17_using_pixel_data.md:299]: No, pages are only marked dirty if they're edited.

Comment (hsautessella) [reply documentation/smg/17_using_pixel_data.md:299]: Not entirely sure what you mean by the second question

Comment (oerc0122) [reply documentation/smg/17_using_pixel_data.md:299]: It looked like you were saying dirty was set on calling `advance`. If you set "dirty" on anything that modifies pixels (presumably the pixels' setter) then that would avoid that.

**Review by oerc0122 — COMMENTED** [review]
Looks good, but I'll wait until we've had the meeting tomorrow to approve.

File: documentation/smg/17_pixel_data.md
Comment (cmarooney-stfc) [inline documentation/smg/17_pixel_data.md:371]: Sorry, but I do not understand this. Vaguely it appears you are getting img_chunk out of sigvar_obj.s as the values for the current page. I do not understand why this is not added directly to obj.signal. I do not understand why img_chunk needs to be put through repelem. I do not understand what smearing means. It would be good if you could provide more explicit explanation here. Thanks

Comment (hsautessella) [reply documentation/smg/17_pixel_data.md:397]: I've added more detail to the problem we're trying to solve in the function header. Hopefully that makes it more understandable, but let me know if it needs further clarification.

File: documentation/smg/17_pixel_data.md
Comment (nickbattam-tessella) [inline documentation/smg/17_pixel_data.md:265]: which function?

File: documentation/smg/17_pixel_data.md
Comment (nickbattam-tessella) [inline documentation/smg/17_pixel_data.md:421]: how? is there a method to do that update?

File: documentation/smg/17_pixel_data.md
Comment (nickbattam-tessella) [inline documentation/smg/17_pixel_data.md:93]: is there a default page size? is there a property containing the page size to set in the object?

File: documentation/smg/17_pixel_data.md
Comment (nickbattam-tessella) [inline documentation/smg/17_pixel_data.md:97]: It would be helpful to include information on how to save an SQW file to merge temp files back and have an updated file on disk.
Does that work if you save back over the original file or does it require a new file to be created?

File: documentation/smg/17_pixel_data.md
Comment (hsautessella) [reply documentation/smg/17_pixel_data.md:265]: I was reluctant to put the name of the function here, since it's a private method and the name could change. If you think it would be better to include the name, I'm happy to do so

**Review by nickbattam-tessella — APPROVED** [review]

**Review by oerc0122 — APPROVED** [review]

File: documentation/smg/17_pixel_data.md
Comment (cmarooney-stfc) [inline documentation/smg/17_pixel_data.md:202]: Not sure this is right? If end==100, then end-10==90 and the range 90:100 has size 11.

File: documentation/smg/17_pixel_data.md
Comment (cmarooney-stfc) [inline documentation/smg/17_pixel_data.md:246]: Instead of multiple what?

File: documentation/smg/17_pixel_data.md
Comment (hsautessella) [reply documentation/smg/17_pixel_data.md]: `directory` was a typo, it was meant to say `directly`.

I've added an example of what's being described here. Hopefully that addresses any confusion.

**Review by oerc0122 — APPROVED** [review]

**Review by cmarooney-stfc — APPROVED** [review]

**Pull request merged**
